### ISSUE TITLE
Issue with `EvictDynamicKey was provided but not was provided any DynamicKey`

### ISSUE THREAD

**Issue by hick209** [VictorAlbertos/RxCache#29]
The first one, throws errors like this one:
`EvictDynamicKey was provided but not was provided any DynamicKey`

To be more specific, this is the stacktrace:
```
Fatal Exception: java.lang.IllegalArgumentException
matchesCacheInfo EvictDynamicKey was provided but not was provided any DynamicKey
 Raw
io.rx_cache.internal.ProxyTranslator.checkIntegrityConfiguration (ProxyTranslator.java:133)
io.rx_cache.internal.ProxyTranslator.processMethod (ProxyTranslator.java:45)
io.rx_cache.internal.ProxyProviders.invoke (ProxyProviders.java:78)
$Proxy1.getChatMessages (Unknown Source)
com.wwn.kickoff.data.repository.ChatDataRepository.getMergedMessages (ChatDataRepository.java:176)
```

It has a reference to `matchesCacheInfo`, but I'm not really calling this method.
Here's the code inside the method `ChatDataRepository.getMergedMessages()` with the line referenced by the stacktrace.

```
    private Observable<CacheData<ChatList>> getMergedMessages(final String username, final long userId, boolean refresh)
    {
        ...

        final Observable<ChatList> localCacheCall = cacheProvider.getChatMessages(Observable.just(new ChatList()), new DynamicKeyGroup(userId, KEY_CHAT_UNSENT_MESSAGES + "-" + username), new EvictDynamicKeyGroup(false));

        ...
    }
```

And here's the `CacheProvider`

```
public interface CacheProvider
{
    ...

    ///////////////////////////////
    /// Matches
    ///////////////////////////////

    Observable<MatchesList> getMatches(Observable<MatchesList> matches, DynamicKey key, EvictDynamicKey evict);
    Observable<CacheData.Info> matchesCacheInfo(Observable<CacheData.Info> cacheInfo, DynamicKey key, EvictDynamicKey evict);

    ...

    ///////////////////////////////
    /// Chat
    ///////////////////////////////

    Observable<ChatList> getChatMessages(Observable<ChatList> messages, DynamicKeyGroup key, EvictDynamicKey evict);
    Observable<CacheData.Info> chatCacheInfo(Observable<CacheData.Info> cacheInfo, DynamicKeyGroup key, EvictDynamicKeyGroup evict);

    @LifeCache(duration = 3, timeUnit = TimeUnit.DAYS)
    Observable<String> getChatMessageDraft(Observable<String> messages, DynamicKeyGroup key, EvictDynamicKeyGroup evict);

    ...
}
```

As you can see, it refers to the method `matchesCacheInfo()` in the stack trace, but the call itself has **nothing** to do with that method.

I've investigated it a fair bit, but still have no clue about what is happening.
Also, it's not a consistent thing, it happens occasionally, not every time.

**Comment by VictorAlbertos** [comment]
Hi @hick209 

It seems like if the Proxy instance which RxCache uses internally would be messing up with the provider being called. 

It would help me a lot if you could narrow the issue with an specific android app posted on github. Even if it crashes only eventually, it would help me a lot. 

Thanks.

**Comment by hick209** [comment]
It's a private repo here. I can't really share the code, but we do have a production version of the app in the PlayStore (https://play.google.com/store/apps/details?id=com.wwn.kickoff) if it helps.

Anything else I could do to help on squasing this potential bug?

**Comment by VictorAlbertos** [comment]
Fells good, seeing other apps in production state which use RxCache :)

Well, you can create a new project from scratch, with minimal code in order to expose the issue. Something like a little repo here on github which narrows the problem to its minimal expression. I know it is hard, but it would help me a lot.

Nevertheless, if you are not able to create the suggested projected, I'll try to replicate the issue by my own next week. And I let you know then.

**Comment by VictorAlbertos** [comment]
Hi @hick209 

I've been digging into the issue, but I wasn't able to catch the potential bug with a set of test. 

But I realised that in the class `ProxyProvider` the instantiation of `ConfigProvider` (which is responsable for throwing that exception) was done before calling the subscription of the observable. So I've deferred the instantiation until the subscription is performed. I think it could be the problem, but I'm not sure. In any case, the exception will be thrown now inside the pipeline of the observable, which is already an improvement :)

Try the new release (I've renamed the packages and bring into this repo the compiler repo for the actionable api, so take a look at the [README file](https://github.com/VictorAlbertos/RxCache) to see how to configure the dependencies using gradle).

`compile "com.github.VictorAlbertos.RxCache:core:1.4.6"`
